tf_quant_finance/black_scholes/approximations/american_option.py

```diff
@@ -0,0 +1,424 @@
+  call option values with the suffix _put and _call.
+
+  ## References:
+  [1] Baron-Adesi, Whaley, Efficient Analytic Approximation of American Option
+    Values, The Journal of Finance, Vol XLII, No. 2, June 1987
+    https://deriscope.com/docs/Barone_Adesi_Whaley_1987.pdf
+
+  Args:
+  volatilities: Real `Tensor` of any shape and dtype. The volatilities to
+    expiry of the options to price.
+  strikes: A real `Tensor` of the same dtype and compatible shape as
+    `volatilities`. The strikes of the options to be priced.
+  expiries: A real `Tensor` of same dtype and compatible shape as
+    `volatilities`. The expiry of each option. The units should be such that
+    `expiry * volatility**2` is dimensionless.
+  spots: A real `Tensor` of any shape that broadcasts to the shape of the
+    `volatilities`. The current spot price of the underlying. Either this
+    argument or the `forwards` (but not both) must be supplied.
+  forwards: A real `Tensor` of any shape that broadcasts to the shape of
+    `volatilities`. The forwards to maturity. Either this argument or the
+    `spots` must be supplied but both must not be supplied.
+  risk_free_rates: An real `Tensor` of same dtype and compatible shape as
+    `volatilities`. Discount factors are calculated using it: e^(-rT), where r
+    is the risk free rate.
+  cost_of_carries: An real `Tensor` of same dtype and compatible shape as
+    `volatilities`. If `spots` is supplied, used to calculate forwards from
+    spots: F = e^(bT) * S.
+  is_call_options: A boolean `Tensor` of a shape compatible with
+    `volatilities`. Indicates whether the option is a call (if True) or a put
+    (if False). If not supplied, call options are assumed.
+  dtype: Optional `tf.DType`. If supplied, the dtype to be used for conversion
+    of any supplied non-`Tensor` arguments to `Tensor`.
+    Default value: None which maps to the default dtype inferred by TensorFlow
+      (float32).
+  name: str. The name for the ops created by this function.
+    Default value: None which is mapped to the default name `option_price`.
+
+  Returns:
+    option_prices: A `Tensor` of the same shape as `forwards`. The Black
+    Scholes price of the options.
+
+  Raises:
+    ValueError: If both `forwards` and `spots` are supplied or if neither is
+      supplied.
+  """
+  if (spots is None) == (forwards is None):
+    raise ValueError('Either spots or forwards must be supplied but not both.')
+
+  # if b>=r, return European only, if b<r, return American only, if mixed, split
+  return tf.case(
+    [(tf.reduce_all(cost_of_carries >= risk_free_rates),
+      lambda: tff.black_scholes.option_price(
+        volatilities, strikes, expiries, spots=spots, forwards=forwards,
+        risk_free_rates=risk_free_rates, cost_of_carries=cost_of_carries,
+        is_call_options=is_call_options, dtype=dtype, name=name)),
+     (tf.reduce_all(cost_of_carries < risk_free_rates),
```

Comment:
Now I see why you didn't like that. Let's for now remove the branching `tf.case` and leave only `split_out` part.
Even `split_out` has too much branching. We need to have access to `_option_price` function itself. Tbh, `gather_nd` is a slow op, and we might as well be faster simply computing `_option_price` and `black_scholes.option_price`. But this would require some benchmarking. For now, should  we add a flag, say, 'use_splitting', where you desribe what you are doing? Any thoughts?

Reply:
I think is actually might be nicer to keep it all in one for now, for the first iteration. Thinking about it, an extra iteration in approximating s_crit would have just the same weight. Take a look now and tell me what you think? I calculate eu prices for all, but american prices only for the ones that need that.